src/sqlfluff/dialects/dialect_ansi.py

```diff
@@ -2304,6 +2307,114 @@ class InsertStatementSegment(BaseSegment):
     )
 
 
+@ansi_dialect.segment()
+class MergeStatementSegment(BaseSegment):
+    """A `MERGE` statement."""
+
+    type = "merge_statement"
+    match_grammar = StartsWith(
+        Ref("MergeIntoLiteralGrammar"),
+    )
+    # Note separate `match_grammar` as overridden in other dialects.
+    parse_grammar = Sequence(
+        Ref("MergeIntoLiteralGrammar"),
+        Indent,
+        OneOf(
+            Ref("TableReferenceSegment"),
+            Ref("AliasedTableReferenceGrammar"),
+        ),
+        Dedent,
+        "USING",
+        Indent,
+        OneOf(
+            Ref("TableReferenceSegment"),
+            Ref("AliasedTableReferenceGrammar"),
+            Sequence(
+                Bracketed(
+                    Ref("SelectableGrammar"),
+                ),
+                Ref("AliasExpressionSegment"),
+            ),
+        ),
+        Dedent,
+        Ref("JoinOnConditionSegment"),
+        Ref("MergeMatchSegment"),
+    )
+
+
+@ansi_dialect.segment()
+class MergeMatchSegment(BaseSegment):
+    """Contains dialect specific merge operations.
+
+    Hookpoint for dialect specific behavior
+    e.g. UpdateClause / DeleteClause, multiple MergeMatchedClauses
+    """
+
+    type = "merge_match"
+    match_grammar = Sequence(
+        Ref("MergeMatchedClauseSegment"),
+        Ref("MergeNotMatchedClauseSegment"),
+    )
+
+
+@ansi_dialect.segment()
+class MergeMatchedClauseSegment(BaseSegment):
+    """The `WHEN MATCHED` clause within a `MERGE` statement."""
+
+    type = "merge_when_matched_clause"
+    match_grammar = Sequence(
+        "WHEN",
+        "MATCHED",
```

Comment:
Postgres, Snowflake, BigQuery all allow an Optional `AND` clause here (and similarly in NOT MATCHED). Could we add this (plus test cases)?

I think that would be also make this complete for BigQuery too for the examples here: https://cloud.google.com/bigquery/docs/reference/standard-sql/dml-syntax#merge_examples so if we could add those examples as fixtures?

Reply:
I agree and have two futher suggestions, that are technically not Ansi but are supported by multiple dialects:
 - ```WHEN MATCHED DELETE``` (Postgres, Snowflake, Spark3, BigQuery)
 - multiple ```WHEN MATCHED / WHEN NOT MATCHED``` (Snowflake, Spark3, BigQuery)

WDYT?

Futhermore there are cases like Postgres ```WHEN NOT MATCHED DO NOTHING``` (and probably plenty more) that I will ignore since adding it to Ansi isn't right.

Reply:
Yeah, ANSI isn't intended to be ANSI:

https://docs.sqlfluff.com/en/stable/dialects.html#ansi
> This dialect doesn’t intend to be brutal in adhering to (and only to) the ANSI SQL spec (mostly because ANSI charges for access to that spec). It aims to be a representation of vanilla SQL before any other project adds their spin to it, and so may contain a slightly wider set of functions than actually available in true ANSI SQL.

So a bit of a judgement call as to what is common enough to put in ANSI (and then override in the **one or two** dialects that don't support that syntax) versus leaving out of ANSI (and then having to override it in **multiple** other dialects that do support this).

So we have flexibility here.

Reply:
I added the following three things to the ANSI Dialect:
 - ```WHEN MATCHED DELETE``` (Postgres, Snowflake, Spark3, BigQuery)
 - multiple ```WHEN MATCHED / WHEN NOT MATCHED``` (Snowflake, Spark3, BigQuery)
 - Optional ```AND``` Clause for ```WHEN MATCHED / WHEN NOT MATCHED```

Reply:
Sounds good. Thanks for that. Will give it one final review (hopefully) when tests are completed and then merge.

Reply:
No worries, thanks for guiding me!